[W5.11][W12-3] Lee Kok Teng

lekoook opened the pull request:
This is for a review

lekoook committed:
Add Sort Command

Sort Command is responsible for sorting the list of contacts in
lexicographical order.

The format of command is: sort

Sort command does nothing for now. Only displays a successful message.

lekoook committed:
Add Sort Command

The sort command now lists the users and displays successful message.

lekoook committed:
Add Sort Command

The sort command now lists the users and displays successful message.

lekoook committed:
Add sort contacts function to SortCommand class

Right now the sort functions only sorts a copy of the current internal list of all contacts

Added Comparator variable to Person class
This comparator variable compares the names of two Person objects and sort them lexicographically in ascending order

Added a getter method to return the underlying internal list in the UniquePersonList class

lekoook committed:
Define function to sort contacts

The sortContacts function in SortCommand class now actually sorts the underlying contact list

Add replaceAllPersons function in AddressBook class to modify the underlying contact list

lekoook committed:
Add documentation for sort command

lekoook committed:
Correct UserGuide format for "Sorting all persons" section

lekoook committed:
Add I/O test cases for sort command

CT15 commented on src/seedu/addressbook/commands/SortCommand.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.List;
+
+/**
+ * Sort the list of contacts in address book
```
- "Sorts" instead of "Sort"
- Punctuation mark required

CT15 commented on src/seedu/addressbook/parser/Parser.java:
```diff
@@ -11,17 +11,7 @@
-import seedu.addressbook.commands.ListCommand;
-import seedu.addressbook.commands.ViewAllCommand;
-import seedu.addressbook.commands.ViewCommand;
+import seedu.addressbook.commands.*;
```
Imported classes should always be listed explicitly.
Configure the appropriate Intellij import settings to avoid this problem.

CT15 commented on src/seedu/addressbook/commands/SortCommand.java:
```diff
@@ -0,0 +1,50 @@
+public class SortCommand extends Command {
+
+    public static final String COMMAND_WORD = "sort";
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Sorts the list of contacts in address book "
```
You need to update `HelpCommand` to show this.

CT15 commented on src/seedu/addressbook/data/AddressBook.java:
```diff
@@ -54,6 +54,10 @@ public void removePerson(ReadOnlyPerson toRemove) throws PersonNotFoundException
         allPersons.remove(toRemove);
     }
 
+    public void replaceAllPersons(UniquePersonList newList) {
```
Instead of changing the entire `UniquePersonList`, you can sort the `internalList` of `UniquePersonList` class.

CT15 commented on src/seedu/addressbook/data/person/Person.java:
```diff
@@ -19,6 +21,21 @@
+     * Comparator to compare two persons based on their names
+     * Lexicographical comparison in ascending order
+     */
+    public static Comparator<Person> personNameComparator = new Comparator<Person>() {
```
This is one way of doing it. Another way will be to make `Person` class implement `Comparable`. To me, implementing `Comparable` will be better as I will consider the lexicographical order the natural ordering of the Person class. Implementing `Comparable` allows you to use `Collections.sort(personList)` immediately.

CT15 commented on src/seedu/addressbook/commands/SortCommand.java:
```diff
@@ -0,0 +1,50 @@
+     *
+     * @param uniquePersonList
+     */
+    private void sortContacts(UniquePersonList uniquePersonList) {
```
Instead of taking out the list, sort this list, create a `UniquePersonList` object with the new sorted list and then replace `allPerson`, you can sort the `internalList` of `allPerson` immediately.

This can be achieved by creating a new method for `UniquePersonList` called, let's say, `sort()`. The implementation looks something like

```java
public void sort() {
    Collections.sort(internalList);  // if Person implements Comparable
}
```

CT15 reviewed (commented):
Great job for this week's enhancement. :)
If you have any questions you can reply to the specific comments.
Otherwise, please close this PR.

lekoook committed:
Improve the description of SortCommand

lekoook committed:
Merge branch 'W5.11' of https://github.com/lekoook/addressbook-level2 into W5.11

lekoook committed:
Fix import wildcard

All imported classes are listed explicitly now.

lekoook committed:
Fix help command

The help command will now display the help text for the sort command

lekoook committed:
Fix sorting methodalogy

Made Person class to implement Comparable so the each Person is compared
lexicographical order based on their names.

Modified the UniquePersonList class by adding a sort method to directly
sort the underlying internal list.

Modified SortCommand class to access the sort command in UniquePersonList
class to sort the underlying list when a user input a sort command.

lekoook commented:
Hi Calvin, thanks for the valuable suggestions!

I have followed them and fixed the areas where needed. Hopefully now its an improved version.
I was initially confused with how Java references work since I was very used to C/C++ pointers to manipulate parameter and return variables. I think I got better understanding now.
